taqueria-protocol/types-config-files.ts

```diff
@@ -0,0 +1,343 @@
+		const configFileObj = await readJsonFile(configFilePath);
+
+		if ((configFileObj as ConfigFileV2).version !== `v2`) {
+			// v1 - only file to load
+			const configFileSetV2 = transformConfigFileV1ToConfigFileSetV2(configFileObj as ConfigFileV1);
+			return configFileSetV2;
+		}
+
+		// Load env files
+		const configFileV2 = configFileObj as ConfigFileV2;
+		const envFiles = await Promise.all(
+			Object.keys(configFileV2.environments ?? {}).map(async envName => {
+				try {
+					return {
+						key: envName,
+						value: await readJsonFile(
+							configFilePath.replace(`config.json`, `config.local.${envName}.json`),
+						) as ConfigEnvironmentFileV2,
+					};
+				} catch {
+					// ignore if file could not be read (no env file)
+					return {};
+				}
+			}),
+		);
+
+		return {
+			config: configFileV2,
+			environments: Object.fromEntries(envFiles.filter(x => x.value).map(x => [x.key, x.value])),
+		};
+	};
+
+export const writeJsonFileInterceptConfig = (writeJsonFile: (filePath: string) => (data: unknown) => Promise<string>) =>
+	(filePath: string): ((data: unknown) => Promise<string>) => {
+		if (filePath.endsWith(`.taq/config.json`)) {
+			return ((config: Config) => {
+				return writeConfigFiles(writeJsonFile)(filePath)(transformConfigToConfigFileV2(config));
+			}) as (data: unknown) => Promise<string>;
+		}
+
+		return writeJsonFile(filePath);
+	};
+
+export const writeConfigFiles = (writeJsonFile: (filePath: string) => (data: unknown) => Promise<string>) =>
+	(configFilePath: string) =>
+		async (configFileSetV2: ConfigFileSetV2) => {
+			const configFileResult = await writeJsonFile(configFilePath)(configFileSetV2.config);
+
+			// write the env files
+			await Promise.all(
+				Object.entries(configFileSetV2.environments).map(async ([envName, value]) => {
+					await writeJsonFile(configFilePath.replace(`config.json`, `config.local.${envName}.json`))(value);
+				}),
+			);
+
+			return configFileResult;
+		};
+
+const removeUndefinedFields = <T>(x: T): T => {
+	return JSON.parse(JSON.stringify(x)) as T;
+};
+
+/** Migrate FileV1 to FileV2
+ *
+ * NOTE: Although this is nearly identical to transformConfigToConfigFileV2
+ * This function should be sealed while the transformConfigToConfigFileV2
+ * will change interatively to become move like V2
+ */
+const transformConfigFileV1ToConfigFileSetV2 = (configFileV1: ConfigFileV1): ConfigFileSetV2 => {
+	const config = configFileV1;
+	const configFileV2: ConfigFileV2 = {
+		version: `v2`,
+		language: config.language,
+		metadata: config.metadata,
+		artifactsDir: config.artifactsDir,
+		contractsDir: config.contractsDir,
+		accounts: !config.accounts
+			? undefined
+			: Object.fromEntries(
+				Object.entries(config.accounts)
+					.map(([k, v]) => [k, { balance: { amount: v, units: `mutez` } }]),
+			),
+		contracts: config.contracts,
+		environmentDefault: config.environment?.default as string,
+		environments: Object.fromEntries(
+			Object.entries(config.environment ?? {})
```

Comment:
If a user has a network and sandbox, but only one env is refering to both, they will lose the second.
One solution will be to make a flat array of environments and their content (sandboxes and networks), so that if an env has 2 sandboxes and 3 networks, 5 copies of that env is created. Then we remove duplicates, then we convert it to the current format.

This probably is not going to happen for any user anyway, so the current solution is perfectly valid as well.

Reply:
Yeah, this will lose data in that case, but I couldn't find any case where that is actually used (other than testing). Let me know if you are aware of any config.json in the wild that might be like that.

Reply:
No, as I said, it's just speculation.